### PR TITLE
GPU Aggregation (5/8): GridLayer

### PR DESCRIPTION
**Pull request by Pessimistress**
For https://github.com/visgl/deck.gl/issues/7457

There is a breaking change in the calculation of bins (division is moved from world space to common space)

#### Change List
- Add `onUpdate` to AggregatorProps
- Add `customOperations` to CPUAggregatorProps for legacy CPUGridLayer `getColorValue`, `getElevationValue` callbacks
- Move GridLayer to use new AggregationLayer and aggregator classes
- Restore unit tests
- Restore some render tests - golden image are updated due to the breaking change above

#### TODO
- The CPUGridLayer-only `*ScaleType`, `*UpperPercentile`, `*LowerPercentile` props are temporarily disabled. This is currently unsupported by the unified Aggregator interface. I think non-linear scales (e.g. step, sqrt, log etc.) would be a nice-to-have for both CPU and GPU aggregation. However, calculating percentiles on GPU is out of scope of v9.1.
- 64-bit support is unavailable. Minor precision differences between CPU and GPU aggregation are expected.

This PR is already pretty big, so I'll try to bring these functionalities back in a separate PR.

### CODE REVIEW

File: modules/aggregation-layers/src/aggregation-layer-v9/cpu-aggregator/aggregate.ts [hunk 1]
```diff
@@ -67,7 +70,7 @@ export function aggregate({
   /** Given the index of a data point, returns its value */
   getValue: (index: number) => number;
   /** Method used to reduce a list of values to one number */
-  operation: AggregationOperation;
+  operation: AggregationOperation | AggregationFunc;
```
Comment (ibgreen) [inline modules/aggregation-layers/src/aggregation-layer-v9/cpu-aggregator/aggregate.ts:73]: Nit: Could make types and code a little cleaner (remove overloading / dynamic type checks) by adding `AggregationOperation: 'CUSTOM'` and supplying the func in a separate `aggregationFunction`... prop?

Perhaps the builtin modes just set a pre-defined aggregationFunction which would read quite easily?

File: modules/aggregation-layers/src/grid-layer/grid-cell-layer-vertex.glsl.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
+export default `#version 300 es
```
Comment (ibgreen) [inline modules/aggregation-layers/src/grid-layer/grid-cell-layer-vertex.glsl.ts:1]: If you haven't started using the syntax highlight plugins (assuming you are using vscode) I recommend it.

```suggestion
export default /* glsl */ `\
#version 300 es
```

See luma.gl docs: https://luma.gl/docs/developer-guide/editing#shader-syntax-highlighting

File: modules/aggregation-layers/src/grid-layer/grid-cell-layer.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+import {defaultColorRange, colorRangeToTexture} from '../utils/color-utils';
+import vs from './grid-cell-layer-vertex.glsl';
+
+const defaultProps: DefaultProps<_GridCellLayerProps> = {
+  colorRange: defaultColorRange,
+  elevationRange: [0, 1000]
+};
+
+/** Proprties added by GridCellLayer. */
+export type _GridCellLayerProps = {
```
Comment (ibgreen) [inline modules/aggregation-layers/src/grid-layer/grid-cell-layer.ts:34]: Nit: Export without underscore in the source file and control underscore in index.ts?

File: modules/aggregation-layers/src/grid-layer/grid-layer.ts [hunk 4]
```diff
@@ -1,112 +1,599 @@
...
   };
 
-  initializeState() {
-    this.state = {
-      useGPUAggregation: false // TODO(v9): Re-enable GPU aggregation.
-    };
+  getAggregatorType(): string {
+    const {
+      gpuAggregation,
+      // lowerPercentile,
```
Comment (ibgreen) [inline modules/aggregation-layers/src/grid-layer/grid-layer.ts:270]: Maybe some TODO  or explanation?

File: test/modules/aggregation-layers/grid-layer.spec.ts [hunk 5]
```diff
@@ -57,113 +57,95 @@ test('GridLayer#updates', t => {
...
-        onAfterUpdate({layer, subLayers, spies}) {
+        onAfterUpdate({layer}) {
           t.ok(
-            layer.state.useGPUAggregation === true,
+            layer.state.aggregator instanceof WebGLAggregator,
             'Should use GPU Aggregation (gpuAggregation: true)'
           );
         }
       },
+      // {
```
Comment (ibgreen) [inline test/modules/aggregation-layers/grid-layer.spec.ts:81]: Add a comment?

File: test/modules/aggregation-layers/grid-layer.spec.ts [hunk 6]
```diff
@@ -57,113 +57,95 @@ test('GridLayer#updates', t => {
...
-          colorScaleType: 'ordinal'
-        },
-        onAfterUpdate({layer, subLayers, spies}) {
-          t.ok(
-            layer.state.useGPUAggregation === false,
-            "Should use CPU Aggregation (colorScaleType: 'ordinal')"
-          );
-        }
       }
+      // {
```
Comment (ibgreen) [inline test/modules/aggregation-layers/grid-layer.spec.ts:127]: Ditto

File: test/modules/gpu-grid-layer.spec.ts [hunk 7]
```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2015 - 2019 Uber Technologies, Inc.
```
Comment (ibgreen) [inline test/modules/gpu-grid-layer.spec.ts:1]: Use vis.gl license headers?

File: test/render/test-cases/grid-layer.js [hunk 8]
```diff
@@ -44,147 +44,117 @@ export function getMax(pts, key) {
 }
 
 export default [
+  // {
```
Comment (ibgreen) [inline test/render/test-cases/grid-layer.js:47]: Comment?

**Review by ibgreen — APPROVED** [review]
Nice work!

File: modules/aggregation-layers/src/grid-layer/grid-layer.ts (hunk 4 above)
Comment (Pessimistress) [reply modules/aggregation-layers/src/grid-layer/grid-layer.ts:272]: See PR description

**Pull request merged**